src/cryptography/hazmat/backends/openssl/ec.py

```diff
@@ -148,6 +151,35 @@ def verify(self):
         return True
 
 
+@utils.register_interface(KeyExchangeContext)
+class _ECDHKeyExchangeContext(object):
+    def __init__(self, backend, private_key, algorithm):
+        self._backend = backend
+        self._private_key = private_key
+        self._kdf = algorithm
+
+    def agree(self, public_key):
+        pri_key = self._private_key._ec_key
+        pub_key = public_key._ec_key
+
+        group = self._backend._lib.EC_KEY_get0_group(pri_key)
+        z_len = (self._backend._lib.EC_GROUP_get_degree(group) + 7) // 8
+        assert z_len > 0
+        z_buf = self._backend._ffi.new("char[]", z_len)
+        peer_key = self._backend._lib.EC_KEY_get0_public_key(pub_key)
+
+        r = self._backend._lib.ECDH_compute_key(z_buf, z_len,
+                                                peer_key, pri_key,
+                                                self._backend._ffi.NULL)
+        assert r > 0
+        z = self._backend._ffi.buffer(z_buf)[:z_len]
+
+        if self._kdf is None:
```

Comment:
For regular DH I was planning on leaving the derviation step out of here but could be convinced to add it back. The main issue is handling of the "None" KDF as it can make it less obvious that a particular use (rather than e.g. all use) of the API is unsafe.